### PR TITLE
snippet inject support like <head lang="en">

### PR DESCRIPTION
**Pull request by oliver-zhang**
Closes https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/8734

### CODE REVIEW

File: instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java [hunk 1]
```diff
@@ -8,9 +8,11 @@
 // this is shared by both ServletOutputStream and PrintWriter injection
 public class InjectionState {
   private static final int HEAD_TAG_WRITTEN_FAKE_VALUE = -1;
-  private static final int HEAD_TAG_LENGTH = "<head>".length();
+  private static final String HEAD_PREFIX = "<head";
+  private static final StringBuilder HEAD_PREFIX_BUILDER = new StringBuilder("");
```
Comment (mateuszrzeszutek) [inline instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java:12]: Having a mutable global variable is bound to fail in a multi-threaded scenario. If the instrumented app handles concurrent requests you will run into races and concurrent modifications.

**Comment by siyuniu-ms** [comment]
There are three solutions currently I could think about:
1. doesn't change the logic for the first 5 char in "<head", and instead of return false when doesn't meet ">" right after, do not change the value unless met with another "<" (this one may slow down the speed of write.)
2. Let user passed in the head tag that they wish to use (may cause problem as the user may have different head for each page)
3. Store a list of some common appearing head tag (I searched a little bit, didn't find any major website put other stuff inside <head> tag, wondering if there any official doc about what could be put inside the tag.)

**Comment by oliver-zhang** [comment]
> 3\. Store a list of some common appearing head tag
Maybe we can lut user passed multiple head tag, and use ```<head>``` as  the default

**Comment by oliver-zhang** [comment]
I write a simple implement for let user passed multiple head tag 
Check each tag,  if there is a match ,then snippet will be injected @siyuniu-ms @mateuszrzeszutek

File: instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java (hunk 1 above)
Comment (oliver-zhang) [reply instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java]: Change another way to do that

**Comment by trask** [comment]
@oliver-zhang is it possible to solve this without requiring user configuration? e.g. when scanning, if we find `<head` and we scan for the closing `>`?

**Comment by oliver-zhang** [comment]
> @oliver-zhang is it possible to solve this without requiring user configuration? e.g. when scanning, if we find `<head` and we scan for the closing `>`?

What do you think? Which way is better? @siyuniu-ms

**Comment by siyuniu-ms** [comment]
I believe the Trask solution is better because it would result in fewer bugs and reduce the workload for customers. This would save them time in terms of becoming familiar with this functionality.

File: instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java [hunk 2]
```diff
@@ -8,7 +8,7 @@
 // this is shared by both ServletOutputStream and PrintWriter injection
 public class InjectionState {
   private static final int HEAD_TAG_WRITTEN_FAKE_VALUE = -1;
-  private static final int HEAD_TAG_LENGTH = "<head>".length();
+  private static final int HEAD_TAG_PREFIX_LENGTH = 4;
```
Comment (mateuszrzeszutek) [inline instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java:11]: You could use the length of the string here
```suggestion
  private static final int HEAD_TAG_PREFIX_LENGTH = "<head".length();
```
for a better readability, and...

File: instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java [hunk 3]
```diff
@@ -40,12 +40,12 @@ public boolean processByte(int b) {
     if (isHeadTagWritten()) {
       return false;
     }
-    if (inHeadTag(b)) {
+    if (inHeadTag(b) && headTagBytesSeen < HEAD_TAG_PREFIX_LENGTH) {
```
[the remark this answers is not in the document]
Comment (oliver-zhang) [reply instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java:43]: The headTagBytesSeen must < HEAD_TAG_PREFIX_LENGTH, then we can headTagBytesSeen++, if remove that extra condition, the headTagBytesSeen might  be  > HEAD_TAG_PREFIX_LENGTH

**Comment by oliver-zhang** [comment]
@mateuszrzeszutek Please take a look at this

**Review by mateuszrzeszutek — COMMENTED** [review]
Hey @oliver-zhang , sorry for the delay. Overall it LGTM 👍 

Could you modify the [existing snippet injection tests](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/servlet/servlet-3.0/javaagent/src/test/groovy/AbstractServlet3Test.groovy) to handle that case?

**Comment by oliver-zhang** [comment]
> Hey @oliver-zhang , sorry for the delay. Overall it LGTM 👍
> 
> Could you modify the [existing snippet injection tests](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/servlet/servlet-3.0/javaagent/src/test/groovy/AbstractServlet3Test.groovy) to handle that case?

@mateuszrzeszutek  Already added

File: instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/Servlet3SnippetInjectingResponseWrapper.java [hunk 4]
```diff
@@ -153,7 +153,8 @@ public boolean isContentTypeTextHtml() {
     if (contentType == null) {
       contentType = super.getHeader("content-type");
     }
-    return contentType != null && contentType.startsWith("text/html");
+    return contentType != null
+        && (contentType.startsWith("text/html") || "application/xhtml+xml".equals(contentType));
```
Comment (trask) [inline instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/Servlet3SnippetInjectingResponseWrapper.java:157]: is this valid `Content-Type`? (wondering if we should do startsWith here)

```
application/xhtml+xml; charset=utf-8
```

**Comment by trask** [comment]
> > Hey @oliver-zhang , sorry for the delay. Overall it LGTM 👍
> > Could you modify the [existing snippet injection tests](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/servlet/servlet-3.0/javaagent/src/test/groovy/AbstractServlet3Test.groovy) to handle that case?
> 
> @mateuszrzeszutek Already added

@mateuszrzeszutek @oliver-zhang what do you think of adding these new tests to the unit tests (`SnippetPrintWriterTest.java` and `SnippetServletOutputStreamTest.java`) instead of the "integration tests"? I think we will eventually end up with lots of edge cases for snippet injection

File: instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/Servlet3SnippetInjectingResponseWrapper.java (hunk 4 above)
Comment (oliver-zhang) [reply instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/Servlet3SnippetInjectingResponseWrapper.java:157]: Yes i agree, this change will have better  compatibility

**Comment by oliver-zhang** [comment]
> @mateuszrzeszutek @oliver-zhang what do you think of adding these new tests to the unit tests (`SnippetPrintWriterTest.java` and `SnippetServletOutputStreamTest.java`) instead of the "integration tests"? I think we will eventually end up with lots of edge cases for snippet injection

@mateuszrzeszutek @trask Already added to (```SnippetPrintWriterTest.java``` and ```SnippetServletOutputStreamTest.java```) ,but i don't remove from "integration tests" right now, if needed i will do it later

**Comment by oliver-zhang** [comment]
@mateuszrzeszutek @trask  Please take a look at this, what else needs to be done?

File: instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java:43]: this logic was a bit hard to follow, what do you think of this alternative? https://github.com/open-telemetry/opentelemetry-java-instrumentation/compare/main...trask:opentelemetry-java-instrumentation:servlet-snippet-inject-alt?expand=1

**Comment by trask** [comment]
> @mateuszrzeszutek @trask Already added to (`SnippetPrintWriterTest.java` and `SnippetServletOutputStreamTest.java`) ,but i don't remove from "integration tests" right now, if needed i will do it later

I'm in favor of removing the new integration tests in this PR, they add a lot of extra code and I don't think have much value over the new "unit tests"

File: instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (oliver-zhang) [reply instrumentation/servlet/servlet-common/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/servlet/InjectionState.java]: nice

**Review by trask — APPROVED** [review]
thx @oliver-zhang!

**Comment by oliver-zhang** [comment]
@mateuszrzeszutek  What do you think ?

**Review by mateuszrzeszutek — APPROVED** [review]
Sorry for the delay; LGTM as well 👍

**Review by laurit — APPROVED** [review]
